core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableMergeUnionRule.java

```diff
@@ -0,0 +1,97 @@
+ * limitations under the License.
+ */
+package org.apache.calcite.adapter.enumerable;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelCollation;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Sort;
+import org.apache.calcite.rel.core.Union;
+import org.apache.calcite.rel.logical.LogicalSort;
+import org.apache.calcite.rel.logical.LogicalUnion;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Rule to convert a {@link org.apache.calcite.rel.logical.LogicalSort} on top of a
+ * {@link org.apache.calcite.rel.logical.LogicalUnion} into a {@link EnumerableMergeUnion}.
+ *
+ * @see EnumerableRules#ENUMERABLE_MERGE_UNION_RULE
+ */
+public class EnumerableMergeUnionRule extends RelRule<EnumerableMergeUnionRule.Config> {
+
+  /** Rule configuration. */
+  public interface Config extends RelRule.Config {
+    Config DEFAULT_CONFIG = EMPTY.withDescription("EnumerableMergeUnionRule").withOperandSupplier(
+        b0 -> b0.operand(LogicalSort.class).oneInput(
+            b1 -> b1.operand(LogicalUnion.class).anyInputs())).as(Config.class);
+
+    @Override default EnumerableMergeUnionRule toRule() {
+      return new EnumerableMergeUnionRule(this);
+    }
+  }
+
+  public EnumerableMergeUnionRule(Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Sort sort = call.rel(0);
+    final RelCollation collation = sort.getCollation();
+    if (collation == null || collation.getFieldCollations().isEmpty()) {
+      return;
+    }
+
+    final Union union = call.rel(1);
+    final int unionInputsSize = union.getInputs().size();
+    if (unionInputsSize < 2) {
+      return;
+    }
+
+    // Push down sort limit, if possible.
+    RexNode inputFetch = null;
+    if (sort.fetch != null) {
+      if (sort.offset == null) {
+        inputFetch = sort.fetch;
+      } else if (sort.fetch instanceof RexLiteral && sort.offset instanceof RexLiteral) {
+        inputFetch = call.builder().literal(
+            RexLiteral.intValue(sort.fetch) + RexLiteral.intValue(sort.offset));
+      }
+    }
+
+    final List<RelNode> inputs = new ArrayList<>(unionInputsSize);
+    for (RelNode input : union.getInputs()) {
+      final RelNode newInput = sort.copy(sort.getTraitSet(), input, collation, null, inputFetch);
+      inputs.add(
+          convert(newInput, newInput.getTraitSet().replace(EnumerableConvention.INSTANCE)));
+    }
+
+    RelNode result = EnumerableMergeUnion.create(sort.getCollation(), inputs, union.all);
+
+    // If Sort contained a LIMIT / OFFSET, then put it back as an EnumerableLimit.
+    // The output of the MergeUnion is already sorted, so we do not need a sort anymore.
+    if (sort.offset != null || sort.fetch != null) {
+      result = EnumerableLimit.create(result, sort.offset, sort.fetch);
```

Comment:
I think you will need to set the collation in  EnumerableLimit? Basically the collation need to be passed through?

Reply:
This is already taken care of by E`numerableLimit#create` method, which calls `RelMdCollation.limit(mq, input)` (which just takes the collation from the limit's input).